stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,395 @@
+
+    # to force "inputs.data_type: float32" in the spec (bonus: disables normalization warning in model._predict_setup)
+    test_input = test_input.astype(np.float32)
+
+    # convert test_input to axes_net semantics and shape, also resize if necessary (to adhere to axes_net_div_by)
+    test_input, axes_img, axes_net, axes_net_div_by, *_ = model._predict_setup(
+        img=test_input,
+        axes=test_input_axes,
+        normalizer=None,
+        n_tiles=None,
+        show_tile_progress=False,
+        predict_kwargs={},
+    )
+
+    # normalization axes string and numeric indices
+    axes_norm = set(axes_net).intersection(set(axes_check_and_normalize(test_input_norm_axes, disallowed='S')))
+    axes_norm = "".join(a for a in axes_net if a in axes_norm)  # preserve order of axes_net
+    axes_norm_num = tuple(axes_net.index(a) for a in axes_norm)
+
+    # normalize input image
+    test_input_norm = normalize(test_input, pmin=min_percentile, pmax=max_percentile, axis=axes_norm_num)
+
+    net_axes_in = axes_net.lower()
+    net_axes_out = axes_check_and_normalize(model._axes_out).lower()
+    ndim_tensor = len(net_axes_out) + 1
+
+    input_min_shape = list(axes_net_div_by)
+    input_min_shape[axes_net.index('C')] = model.config.n_channel_in
+    input_step = list(axes_net_div_by)
+    input_step[axes_net.index('C')] = 0
+
+    # the axes strings in bioimageio convention
+    input_axes = "b" + net_axes_in.lower()
+    output_axes = "b" + net_axes_out.lower()
+
+    if mode == "keras_hdf5":
+        output_names = ("prob", "dist") + (("class_prob",) if model._is_multiclass() else ())
+        output_n_channels = (1, model.config.n_rays,) + ((1,) if model._is_multiclass() else ())
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        output_scale = [1]+list(1/g for g in model.config.grid) + [0]
+        output_offset = [0]*(ndim_tensor)
+
+    elif mode == "tensorflow_saved_model_bundle":
+        if model._is_multiclass():
+            raise NotImplementedError("Tensorflow SavedModel not supported for multiclass models yet")
+        # regarding input/output names: https://github.com/CSBDeep/CSBDeep/blob/b0d2f5f344ebe65a9b4c3007f4567fe74268c813/csbdeep/utils/tf.py#L193-L194
+        input_names = ["input"]
+        output_names = ["output"]
+        output_n_channels = (1 + model.config.n_rays,)
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        # same shape as input except for the channel dimension
+        output_scale = [1]*(ndim_tensor)
+        output_scale[output_axes.index("c")] = 0
+        # no offset, except for the input axes, where it is output channel / 2
+        output_offset = [0.0]*(ndim_tensor)
+        output_offset[output_axes.index("c")] = output_n_channels[0] / 2.0
```

Comment:
This seems rather unintuitive to me, i.e. specifying an offset to set the number of output channels (?)

Reply:
We have currently to options for doing this (when using a shape description relative to the input shape):
1. specifying it as a `scale` factor that is multiplied by the input shape
2. specfiying it as an `offset` factor that is added

This can also be combined, so the full equation is `output_shape = scale * output_shape + 2 * offset`. It may indeed be a bit confusing, because this was done with spatial axes in mind. But we don't want to introduce new spec changes any time soon in order to not derail the publication process. So our two options here are:
1. `scale["c"] = output_channels / input_channels; offset["c"] = 0`
2. `scale["c"] = 0; offset["c"] = output_channels / 2`

I chose option 2 because I found it less confusing. But we could also change to option 1 if you think it's easier to understand. And certainly add a comment to explain it better.